Auth Map: Initial Garbage Collection

mhofstetter opened the pull request:
This PR introduces garbage collection for the auth map based on the following events:

* deleted cilium node: all auth map entries which belong to the deleted node will be deleted
* deleted cilium identity: all auth map entries which belong to the deleted identity will be deleted
* timer (configurable - defaults to every 15m): expired auth map entries will be deleted

The implementation uses the [hive job framework](https://docs.cilium.io/en/latest/contributing/development/hive/#job-groups).

Known Limitations:

* Garbage Collection based on deleted Cilium Identities is currently only supported if Cilium Identities are backed by the CRD storage. Eventually the expired entries will be cleaned up by the periodic timer based run.
* Currently no GC based on deleted policies
 
Fixes: #25213

mhofstetter committed:
auth: provide (all) CiliumNodes and identity privately

Currently, only the local CiliumNode is available in the shared
resources.

This commit introduces the possibility to watch all CiliumNodes
and CiliumIdentities within the auth cell by providing them privately
for the cell only.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
auth: implementation of auth map garbage collection

This commit introduces garbage collection for the auth map based on the
following events:

* deleted node: all auth map entries which belong to the deleted node
  will be deleted
* deleted identity: all auth map entries which belong to the deleted
  identity will be deleted
* timer (configurable - defaults to every 15m): expired auth map entries
  will be deleted

Garbage Collection based on deleted Cilium Identities is currently only
supported if Cilium Identities are backed by the CRD storage.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

jrajahalme reviewed (commented):
Looking good :-)

You may need to tap into `func (d *Daemon) UpdateIdentities(added, deleted cache.IdentityCache)` to get identity deletes regardless of identity allocation mode.

jrajahalme commented on pkg/auth/authmap_gc.go:
```diff
@@ -0,0 +1,203 @@
+}
+
+func (r *authMapGarbageCollector) cleanupEntries(predicate func(key authKey, info authInfo) bool) error {
+	all, err := r.authmap.All()
```
Accessing the map returned by `All()` seems risky, as authmap could modify the map concurrently?

mhofstetter replied:
@jrajahalme thanks! replaced method `cleanupEntries`in GC with `DeleteIf` in the cache itself. this allows to lock during this operation.

jrajahalme commented on pkg/auth/authmap_gc.go:
```diff
@@ -0,0 +1,203 @@
+	}
+
+	var entriesToDelete []authKey
+	for k, v := range all {
```
IMO this iteration needs to be done with the map locked?

jrajahalme replied:
If locking the `all` map becomes too heavy, then maybe collect "gc event info" and iterate through the bpf map once per GC period, removing all the needed entries (expired, removed IDs, removed nodes) during one iteration?

mhofstetter replied:
@jrajahalme agree. replaced method cleanupEntriesin GC with DeleteIf in the cache itself. this allows to lock during this operation.

thanks for the proposal with keeping the events. agree that this might be an option if the lock becomes too heavy - but i think we can go with the direct GC for the time being?

viktor-kurchenko reviewed (approved):
LGTM.

jrajahalme reviewed (approved):
`authMapCache.All()` is still returning the internal map to be used with the internal lock released. This function does not seem to be used however, so maybe split the interface requiring `All()` to be implemented into two, one with it and one without? Or make it to return a copy of the map instead?

mhofstetter committed:
auth: authmapCache.All() returns copy of map

This commit lets authmapCache.All() returning a copy of the map of
cached entries instead of the map instead. This is necessary to prevent
issues when using the map without acquiring the lock.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

dylandreimerink commented on pkg/auth/authmap_gc.go:
```diff
@@ -0,0 +1,174 @@
+		if r.discoveredCiliumIdentities != nil {
+			log.Debug("auth: identities discovered")
+			id, err := identity.ParseNumericIdentity(e.Object.Name)
+			if err != nil {
+				return fmt.Errorf("failed to parse identity: %w", err)
+			}
```
This branch is missing the e.Done call. I recommend to restructure this and the other handler, make a `err` variable at the start of the function, defer the `e.Done(err)` then set `err` in all of your branches, that way you can't ever forget to call `e.Done`

joamaki replied:
Bit of background: `e.Done` marks the event as handled and without the call we'd never process an event for the same key again. `pkg/k8s/resource` guards against this with a finalizer function that will panic if we ever GC an event that wasn't marked done.

mhofstetter replied:
@dylandreimerink thanks - good catch! refactored it towards your approach!

joamaki commented on daemon/cmd/cells.go:
```diff
@@ -37,6 +38,9 @@ var (
 		ControlPlane,
 		datapath.Cell,
+
+		// Provides a global job registry which cells can use to spawn job groups.
```
This should be part of the `Infrastructure` module. Hmm or even just part of `ControlPlane` as otherwise we'd need to repeat it in e.g. `test/controlplane/suite/agent.go` and I'd like to avoid having too many exceptions there. Though even that feels very wrong as the jobs would be something that both `Infrastructure` and `Datapath` would want.

@dylandreimerink or alternatively we'll just make this part of the "defaults" provided by Hive? Might be cleanest.

dylandreimerink replied:
> or alternatively we'll just make this part of the "defaults" provided by Hive? Might be cleanest.

I like that approach, given that it works so well with the life-cycles which are also provided by default. But is that a change that belongs in this PR?

mhofstetter replied:
@joamaki @dylandreimerink thanks! will leave this as is. AFAIU this will be provided via the defaults in an upcoming PR by @dylandreimerink

mhofstetter replied:
job cell registration has been removed from this PR as this is already done with https://github.com/cilium/cilium/pull/25471 in infrastructure

dylandreimerink reviewed (changes requested):
Nice work! Just the one thing

joamaki commented on pkg/auth/authmap_gc.go:
```diff
@@ -0,0 +1,174 @@
+	}
+}
+
+func (r *authMapGarbageCollector) HandleCiliumNodeEvent(_ context.Context, e resource.Event[*ciliumv2.CiliumNode]) error {
```
I wonder if this should be hooking into `NodeManager` instead of subscribing to `Resource[*CiliumNode]` directly? Note that `Resource[*CiliumNode]` won't contain nodes from other clusters...

mhofstetter replied:
@joamaki thanks for the hint. AFAIK Mutual Auth in 1.14 won't be compatible with ClusterMesh. Therefore, this should be ok. But i will create an issue to track this!

joamaki commented on pkg/auth/authmap_gc.go:
```diff
@@ -0,0 +1,174 @@
+	return nil
+}
+
+func (r *authMapGarbageCollector) HandleCiliumIdentityEvent(_ context.Context, e resource.Event[*ciliumv2.CiliumIdentity]) error {
```
`CiliumIdentity` is only valid when identity mode is CRD. This won't work with kvstore.

mhofstetter replied:
@joamaki yes, this is mentioned in the PR description. will be tracked and refactored towards an abstraction which covers kvstore & crd based backend.

btw: as discussed offline - resource cells cilium nodes & cilium identity are now provided privately only within the auth cell.

joamaki commented on pkg/auth/authmap_cache.go:
```diff
@@ -66,6 +73,32 @@ func (r *authMapCache) Delete(key authKey) error {
+		}
+	}
+
+	for _, key := range entriesToDelete {
```
Why wouldn't we just do a single loop that checks the predicate and deletes?

mhofstetter replied:
@joamaki thought it isn't safe to delete entries of the map which gets iterated over the same time.

thanks for the hint that it is! -> fixed

https://go.dev/ref/spec#For_statements

> The iteration order over maps is not specified and is not guaranteed to be the same from one iteration to the next. If a map entry that has not yet been reached is removed during iteration, the corresponding iteration value will not be produced. If a map entry is created during iteration, that entry may be produced during the iteration or may be skipped. The choice may vary for each entry created and from one iteration to the next. If the map is nil, the number of iterations is 0.

joamaki commented on pkg/auth/authmap_gc.go:
```diff
@@ -0,0 +1,175 @@
+
+func (r *authMapGarbageCollector) HandleCiliumNodeEvent(_ context.Context, e resource.Event[*ciliumv2.CiliumNode]) error {
+	var err error
+	defer e.Done(err)
```
This doesn't work. the `err` is evaluated eagerly here and will always be nil. You need to do `func() { e.Done(err) }`.
https://go.dev/play/p/xbVK-puk5cD.

I would do `(err error)` return value so we'll capture the context too (e.g. the "failed to cleanup missing codes"). Consider adding unit tests to verify that the right thing happens. `type Event struct` makes this pretty easy as `Done` is just a field.

@dylandreimerink double-check that this isn't happening in your code as well and that retry handling isn't broken because of this.

mhofstetter replied:
@joamaki :facepalm: of course, thanks. 

fixed it and added unit-tests which are checking the returned and also the reported error (via event done)

joamaki replied:
Awesome thanks!

joamaki commented on pkg/auth/authmap_writer.go:
```diff
@@ -77,3 +80,28 @@ func (r *authMapWriter) Delete(key authKey) error {
+		}
+	}
+
+	for _, key := range entriesToDelete {
```
Same question as earlier. What's the benefit of the more complicated double looping and extra data structure?

mhofstetter replied:
@joamaki thanks - fixed!

joamaki commented on pkg/auth/cell.go:
```diff
@@ -36,16 +45,28 @@ var Cell = cell.Module(
-	cell.Config(config{MeshAuthQueueSize: 1024}),
+	// Providing k8s resource Node & Identity privately to avoid further usage of them in other agent components
+	cell.ProvidePrivate(
+		// TODO: use node manager to get events of all nodes, including the ones of other clusters (ClusterMesh)
```
Can we drop the TODOs and add github issues instead?

mhofstetter replied:
@joamaki done

joamaki commented on pkg/auth/authmap_cache.go:
```diff
@@ -26,7 +29,11 @@ func (r *authMapCache) All() (map[authKey]authInfo, error) {
 	defer r.cacheEntriesMutex.RUnlock()
 
-	return r.cacheEntries, nil
+	result := make(map[authKey]authInfo)
```
https://pkg.go.dev/golang.org/x/exp/maps#Clone might do the same? Either way works for me :)

mhofstetter replied:
@joamaki yes it's doing the same - changed to use it. didn't check whether we're using this extension before.

joamaki commented on pkg/auth/authmap_gc.go:
```diff
@@ -0,0 +1,173 @@
+	}
+}
+
+func (r *authMapGarbageCollector) HandleCiliumNodeEvent(_ context.Context, e resource.Event[*ciliumv2.CiliumNode]) (err error) {
```
Do these methods have to be exported? Seeing it exported immediately makes me think it must be called from outside this package and then wondering why these aren't made to be thread-safe and stuff like that.

mhofstetter replied:
@joamaki unexported :+1:

joamaki commented on pkg/auth/cell.go:
```diff
@@ -97,9 +124,43 @@ func newManager(params authManagerParams) error {
+	if params.CiliumNodes != nil {
+		jobGroup.Add(job.Observer("auth nodes gc",
+			mapGC.HandleCiliumNodeEvent,
+			stream.FromChannel(params.CiliumNodes.Events(context.Background())),
```
I would prefer if this would use a context that is cancelled when this stops. 

Looking at this code I think it'd be better if `Resource[T]` would also be an `Observable`. It used to be, but then I decided to simplify it as in pretty much all the cases we didn't care about completion. Implementation is simple though:

```
diff --git a/pkg/k8s/resource/resource.go b/pkg/k8s/resource/resource.go
index 8f0c75919f..b11f0a9982 100644
--- a/pkg/k8s/resource/resource.go
+++ b/pkg/k8s/resource/resource.go
@@ -17,6 +17,7 @@ import (
        "github.com/cilium/cilium/pkg/hive"
        "github.com/cilium/cilium/pkg/lock"
        "github.com/cilium/cilium/pkg/promise"
+       "github.com/cilium/cilium/pkg/stream"
 )
 
 // Resource provides access to a Kubernetes resource through either
@@ -40,6 +41,8 @@ import (
 // The resource is lazy, e.g. it will not start the informer until a call
 // has been made to Events() or Store().
 type Resource[T k8sRuntime.Object] interface {
+       stream.Observable[Event[T]]
+
        // Events returns a channel of events. Each event must be marked as handled
        // with a call to Done(), otherwise no new events for this key will be emitted.
        //
@@ -295,6 +298,10 @@ func WithErrorHandler(h ErrorHandler) EventsOpt {
        }
 }
 
+func (r *resource[T]) Observe(ctx context.Context, next func(Event[T]), complete func(error)) {
+       stream.FromChannel(r.Events(ctx)).Observe(ctx, next, complete)
+}
+
 // Events subscribes the caller to resource events.
 //
 // Each subscriber has their own queues and can process events at their own
```

Up for doing this, or shall we punt it to a follow-up PR?

mhofstetter replied:
@joamaki will open a follow-up PR. i think making the k8s Resource an observable is the only logical step - otherwise we're back to using a dedicated lifecycle start&stop hook...

mhofstetter replied:
@joamaki i opened https://github.com/cilium/cilium/pull/25934 for this

joamaki commented on pkg/auth/manager_test.go:
```diff
@@ -138,14 +149,33 @@ func (r *fakeAuthHandler) subscribeToRotatedIdentities() <-chan certs.Certificat
+		return errors.New("failed to delete entry")
+	}
+
+	for k, v := range r.entries {
```
Could codegolf this to: `maps.DeleteFunc(r.entries, predicate)` ;-)

mhofstetter replied:
@joamaki changed, thanks
